internal/scheduler/scheduler.go

```diff
@@ -317,10 +326,39 @@ func (w *jobWrapper) Build() (*batchv1.Job, error) {
 			{
 				Name:  "BUILDKITE_AGENT_EXPERIMENT",
 				Value: "kubernetes-exec",
-			}, {
+			},
+			{
 				Name:  "BUILDKITE_CONTAINER_COUNT",
 				Value: strconv.Itoa(containerCount),
 			},
+			{
+				Name:  "BUILDKITE_AGENT_TAGS",
+				Value: createAgentTagString(agentTags),
```

Comment:
... wait, if we're using this env, why not pass the others in here as well?

Reply:
So the env vars are passed as a `valueFrom` so the scheduler does not know their values. These are filled in by some part of Kubernetes by the time the containers in the pod are running.

Here's an excerpt of what the scheduler will output in yaml form:
```yaml
...
    spec:
      containers:
...
	  - name: agent
	    command: buildkite-agent
        args: start
	    env:
        - name: BUILDKITE_K8S_NODE
          valueFrom:
            fieldRef:
              fieldPath: spec.nodeName
...
```

Not only does the scheduler not what know that the value of `BUILDKITE_K8S_NODE` will be at this time, but for this particular example of the node name, it **can't** know because the ultimate decision about which node the pod is run on is made by [kube-scheduler](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/), not this scheduler. (For this reason I prefer to call this code path the "controller" and not the "scheduler", but as the filename is `scheduler.go`, that generally leads to more confusion.)

This controller just calls the Kubernetes API with the definition of the Kubernetes job. The YAML form of that job definition will contain the yaml snippet above.

For `BUILDKITE_K8S_NAMESPACE` and `BUILDKITE_K8S_SERVICE_ACCOUNT`, it is possible to inject them into the `--tags` argument here, but that would require making sure the source of truth for these is respected in this method, which will be fragile. There's less coupling between different parts of this very long method if we just set these once in this method and pick them up at runtime in the same way as the node name.

Reply:
Ahhh, yeah I see. Sorry, I wasn't quite tracing that through in my head. But this makes sense. :+1: